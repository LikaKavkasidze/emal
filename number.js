/// Emal – Acquire and process scientific data with ease
/// number.js – Parser of arbitrary precision numbers.
///
/// An Emal number is stored as a large integer (BigInt) with decimal
/// place position.
/// Parsing is based on a state machine with states corresponding to
/// parts of the number.

import { StateMachine } from "./stateMachine.js";
import { isDigit, isSign, isComma } from "./matchers.js";

// Default decimal count for `toString`
const ENUMBER_DEFAULT_DC = 2;
// Default extra decimals computed for division
const ENUMBER_DEFAULT_DIV_OFFSET = BigInt(ENUMBER_DEFAULT_DC + 2);

// Define a start state which defines a local variable and simply returns.
// This declared local variable is used to construct tokens generically, so 
// that parsing of the number and exponent uses the same states.
function start(charCode) {
	this.isExponent = false;
	return this.sign(charCode);
}

function sign(charCode) {
	if(!isSign(charCode)) {
		return this.startIntPart(charCode);
	}

	this.token({}, `${this.isExponent ? "eS" : "s"}ign`);
	this.eat(charCode);

	return startIntPart;
}

function startIntPart(charCode) {
	if(!isDigit(charCode)) {
		return this.nok(charCode);
	}

	this.token({}, `${this.isExponent ? "eI" : "i"}ntPart`);
	this.eat(charCode);

	return intPart;
}

function intPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	// Exponents cannot have a floating part
	} else if(isComma(charCode) && !this.isExponent) {
		this.token({}, "floatPart");
		return floatPart;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return intPart;
	}

	return this.end(charCode);
}

function floatPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return floatPart;
	}

	return this.end(charCode);
}

// Used as a transformer on token values which are strings
function toNumber(token) {
	return BigInt(token.value);
}

function toValue(token) {
	return token.value;
}

export const numberSm = new StateMachine("start",
	{
		start,
		sign,
		startIntPart,
		intPart,
		floatPart,
	},
	{
		transformers: {
			sign: toValue,
			intPart: toValue,
			floatPart: toValue,
			eSign: toValue,
			eIntPart: toNumber,
		},
		type: "parallel",
	},
);

function BIPow(n, p) {
	let o = n;

	for(; p > 1; p--) {
		o *= n;
	}

	return o;
}

function preprocessOp(a, b) {
	let aInt = a.rawInt;
	let bInt = b.rawInt;

	let aLen = a.decimalPlace;
	let bLen = b.decimalPlace;

	if(aLen > bLen) {
		const delta = aLen - bLen;
		bInt *= BIPow(10n, delta);
		bLen += delta;
	} else if(bLen > aLen) {
		const delta = bLen - aLen;
		aInt *= BIPow(10n, delta);
		aLen += delta;
	}

	return {
		aInt,
		bInt,
		oLen: aLen,
	};
}

export class EmalNumber {
	constructor(rawInt, decimalPlace) {
		this.rawInt = rawInt;
		this.decimalPlace = decimalPlace;
	}

	static fromTokens(userTokens) {
		// Use defaults because some parts of the number are optional
		const defaultTokens = {
			sign: "+",
			floatPart: "0",
			eSign: "+",
			eIntPart: "0",
		};

		const tokens = Object.assign({}, defaultTokens, userTokens);

		const rawInt = BigInt(`${tokens.sign}${tokens.intPart}${tokens.floatPart}`);
		const exponent = BigInt(`${tokens.eSign}${tokens.eIntPart}`);
		const decimalPlace = BigInt(tokens.floatPart.length) - exponent;

		return new EmalNumber(rawInt, decimalPlace);
	}

	static fromString(numberStr) {
		const runnerOutput = numberSm.run(numberStr).transform();
		return EmalNumber.fromTokens(runnerOutput.tokens);
	}

	toString(decimalCount) {
		// Dynamic default argument
		if(typeof decimalCount !== "number") decimalCount = ENUMBER_DEFAULT_DC;

		const { rawInt, decimalPlace } = this;

		const intStr = rawInt.toString(10);
		const exponent = BigInt(intStr.length) - decimalPlace - 1n;

		const intPart = intStr.substr(0, 1);
		const floatPart = intStr.substr(1, decimalCount);

		return `${intPart},${floatPart}${exponent !== 0n ? "e".concat(exponent) : ""}`
	}

	static add(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt + bInt, oLen);
	}

	static sub(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt - bInt, oLen);
	}

	static mul(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt * bInt, oLen * 2n);
	}

	static div(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		// Dividing arbitrary precision numbers cannot be done
		// with infinite precision, here the number is cut to
		// ENUMBER_DEFAULT_DIV_OFFSET digits after decimal point.
		const offset = ENUMBER_DEFAULT_DIV_OFFSET + oLen;
		return new EmalNumber(aInt * BIPow(10n, offset) / bInt, offset);
	}
}
