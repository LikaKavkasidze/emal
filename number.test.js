import { EmalNumber } from "./number.js";

test("tokenizes numbers", () => {
	const simple = EmalNumber.fromString("9.44");
	const scientific = EmalNumber.fromString("4.56e-6");
	const negative = EmalNumber.fromString("-4.5e-3");

	expect(simple).toMatchSnapshot();
	expect(scientific).toMatchSnapshot();
	expect(negative).toMatchSnapshot();
});

test("allows commas", () => {
	const n = EmalNumber.fromString("520,73");
	expect(n).toMatchSnapshot();
});

test("finds length", () => {
	const simple = EmalNumber.fromString("9.44");
	const scientific = EmalNumber.fromString("4.56e-6");
	const negative = EmalNumber.fromString("-4.5e-3");

	expect(simple.rawLength()).toEqual(3n);
	expect(scientific.rawLength()).toEqual(3n);
	expect(negative.rawLength()).toEqual(2n);
});

test("serializes numbers", () => {
	const simpleStr = "9.44";
	const scientificStr = "4.56e-6";
	const negativeStr = "-4.5e-3";

	const simple = EmalNumber.fromString(simpleStr);
	const scientific = EmalNumber.fromString(scientificStr);
	const negative = EmalNumber.fromString(negativeStr);

	expect(simple.toString()).toEqual(simpleStr.replace(".", ","));
	expect(scientific.toString()).toEqual(scientificStr.replace(".", ","));
	expect(negative.toString(1)).toEqual(negativeStr.replace(".", ","));
});

test("rounds correctly", () => {
	const str = "8,5888";
	const n = EmalNumber.fromString(str);
	expect(n.toString()).toEqual("8,59");
});

test("adds numbers", () => {
	const a = EmalNumber.fromString("4.56e-2");
	const b = EmalNumber.fromString("1,07");

	expect(EmalNumber.add(a, b).toString(4)).toEqual("1,1156");
});

test("divides numbers", () => {
	const a = EmalNumber.fromString("4.56e-2");
	const b = EmalNumber.fromString("1,07");

	expect(EmalNumber.div(a, b).toString()).toEqual("4,26e-2");
});

test("log numbers", () => {
	const x = new EmalNumber(23n, 1n);
	const y = new EmalNumber(87n, 2n);

	expect(EmalNumber.log10(x).toString()).toEqual("3,62e-1");
	expect(EmalNumber.log10(y).toString()).toEqual("-6,05e-2");
});
