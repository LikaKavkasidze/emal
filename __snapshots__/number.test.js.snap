// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows commas 1`] = `
EmalNumber {
  "decimalPlace": 2n,
  "isNegative": false,
  "rawInt": 52073n,
}
`;

exports[`tokenizes numbers 1`] = `
EmalNumber {
  "decimalPlace": 2n,
  "isNegative": false,
  "rawInt": 944n,
}
`;

exports[`tokenizes numbers 2`] = `
EmalNumber {
  "decimalPlace": 8n,
  "isNegative": false,
  "rawInt": 456n,
}
`;

exports[`tokenizes numbers 3`] = `
EmalNumber {
  "decimalPlace": 4n,
  "isNegative": true,
  "rawInt": -45n,
}
`;
